x-pack/filebeat/input/internal/private/private_test.go

```diff
@@ -0,0 +1,366 @@
+// Copyright Elasticsearch B.V. and/or licensed to Elasticsearch B.V. under one
+// or more contributor license agreements. Licensed under the Elastic License;
+// you may not use this file except in compliance with the Elastic License.
+
+package private
+
+import (
+	"bytes"
+	"encoding/json"
+	"net/url"
+	"reflect"
+	"testing"
+
+	"github.com/google/go-cmp/cmp"
+)
+
+type redactTest struct {
+	name    string
+	in      any
+	tag     string
+	global  []string
+	want    any
+	wantErr error
+}
+
+var redactTests = []redactTest{
+	{
+		name: "map_string",
+		in: map[string]any{
+			"private":    "secret",
+			"secret":     "1",
+			"not_secret": "2",
+		},
+		want: map[string]any{
+			"private":    "secret",
+			"not_secret": "2",
+		},
+	},
+	{
+		name: "map_string_inner",
+		in: map[string]any{
+			"inner": map[string]any{
+				"private":    "secret",
+				"secret":     "1",
+				"not_secret": "2",
+			}},
+		want: map[string]any{
+			"inner": map[string]any{
+				"private":    "secret",
+				"not_secret": "2",
+			}},
+	},
+	{
+		name: "map_string_inner_global",
+		in: map[string]any{
+			"inner": map[string]any{
+				"secret":     "1",
+				"not_secret": "2",
+			}},
+		global: []string{"inner.secret"},
+		want: map[string]any{
+			"inner": map[string]any{
+				"not_secret": "2",
+			}},
+	},
+	{
+		name: "map_string_inner_next_inner_global",
+		in: map[string]any{
+			"inner": map[string]any{
+				"next_inner": map[string]any{
+					"secret":     "1",
+					"not_secret": "2",
+				},
+			}},
+		global: []string{"inner.next_inner.secret"},
+		want: map[string]any{
+			"inner": map[string]any{
+				"next_inner": map[string]any{
+					"not_secret": "2",
+				},
+			}},
+	},
+	{
+		name: "map_string_inner_next_inner_params_global",
+		in: map[string]any{
+			"inner": map[string]any{
+				"next_inner": map[string]any{
+					"headers": url.Values{
+						"secret":     []string{"1"},
+						"not_secret": []string{"2"},
+					},
+					"not_secret": "2",
+				},
+			}},
+		global: []string{"inner.next_inner.headers.secret"},
+		want: map[string]any{
+			"inner": map[string]any{
+				"next_inner": map[string]any{
+					"headers": url.Values{
+						"not_secret": []string{"2"},
+					},
+					"not_secret": "2",
+				},
+			}},
+	},
+	{
+		name: "map_string_inner_next_inner_params_global_internal",
+		in: map[string]any{
+			"inner": map[string]any{
+				"next_inner": map[string]any{
+					"headers": url.Values{
+						"secret":     []string{"1"},
+						"not_secret": []string{"2"},
+					},
+					"not_secret": "2",
+				},
+			}},
+		global: []string{"inner.next_inner.headers"},
+		want: map[string]any{
+			"inner": map[string]any{
+				"next_inner": map[string]any{
+					"not_secret": "2",
+				},
+			}},
+	},
+	{
+		name: "map_slice",
+		in: map[string]any{
+			"private":    []string{"secret"},
+			"secret":     "1",
+			"not_secret": "2",
+		},
+		want: map[string]any{
+			"private":    []string{"secret"},
```

Comment:
What do you mean? That it should be removed, or that it should be retained? This is configurable; if the slice here contained the string "private", it would also be removed.